### PR TITLE
Feature/onboarding

### PR DESCRIPTION
**Pull request by Uros-Bijelic**
Finished onboarding user feature with storing user into database. Updated middleware.ts file with GET user call, to have property onboardingStep so we can redirect user to finish onboarding process if he logged out before finish. 
Created separate api route  /api/user/ so to fetch user from session.

### CODE REVIEW

File: app/onboarding/page.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  try {
+    user = await getUserById(session.user.id);
+  } catch (error) {
+    console.log('Error fetching ', error);
+  }
+
+  if (!user) {
+    return null;
+  }
```
Comment (JazzyGrim) [inline app/onboarding/page.tsx]:
```suggestion
  const user: IUser | null = await getUserById(session.user.id);
  
  if (!user) {
    return null;
  }
```

We can just fetch the user. If we get an error we don't have to catch it, instead we can create an `error` file which NextJS will use to show an error to the user. This is called an error boundary.

File: components/onboarding/BasicInformations.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  const { LEARNING_GOALS } = EOnboardingStep;
+  const { trigger, setValue, getValues } = useFormContext();
+  const [uploadedImage, setUploadedImage] = useState('');
+
+  const onSuccessUpload = (
+    result: CloudinaryUploadWidgetResults,
+    options: any
+  ) => {
+    console.log('result', result);
+    console.log('options', options);
```
Comment (JazzyGrim) [inline components/onboarding/BasicInformations.tsx]: Leftover

File: components/onboarding/BasicInformations.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    if (!result || !result?.info) throw new Error('Image not uploaded!');
```
Comment (JazzyGrim) [inline components/onboarding/BasicInformations.tsx]:
```suggestion
    if (!result?.info) throw new Error('Image not uploaded!');
```
This should do the same thing

File: components/onboarding/BasicInformations.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    setUploadedImage((result.info as CloudinaryUploadWidgetInfo).secure_url);
```
Comment (JazzyGrim) [inline components/onboarding/BasicInformations.tsx]:
```suggestion
    if (result.info instanceof String) throw new Error('Image not uploaded');
    setUploadedImage(result.info.secure_url);
```
The info can be a string or upload result. If it's a string it's likely an error occured. We can check, and also avoid type casting this way.

File: components/onboarding/OnboardingContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+  const handleChangeStep = async (
+    data: Partial<IUserOnboarding>,
+    newStep: EOnboardingStep
+  ) => {
+    if (!_id) return;
+
+    await updateUserOnboardingStep(user._id.toString(), data);
+    setStep(newStep);
+    console.log('handleChangeStep opalio');
```
Comment (JazzyGrim) [inline components/onboarding/OnboardingContainer.tsx]: leftover

File: components/shared/Stepper.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ * step 2 width 33%
+ * step 3 width 66%
+ * step 4 width 100%
+ *
+ */
+/**
+ * 1.  (1 - 1) / x * 100 = 0%
+ * 2.  (2 - 1) / x  * 100 = 33%
+ * 3.  (3 - 1) / x * 100 = 66%
+ */
```
Comment (JazzyGrim) [inline components/shared/Stepper.tsx]: leftover

File: components/shared/learning-goals/GoalsAndKnowledgeRow.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+// TODO maybe this can be reused, left just in case i need it later
+
+interface ILearningGoalProps {
+  name: string;
+  label: string;
+  handleDelete?: () => void;
+}
+
+const GoalsAndKnowledgeRow: React.FC<ILearningGoalProps> = ({
+  name,
```
Comment (JazzyGrim) [inline components/shared/learning-goals/GoalsAndKnowledgeRow.tsx]: We don't need this file

File: lib/actions/user-actions.ts [hunk 8]
```diff
@@ -56,3 +66,49 @@ export const createNewUser = async ({
...
+
+    const user = await User.findOneAndUpdate<IUser>(
+      { _id: id },
+      { $set: data }
+    ).lean();
+
+    return JSON.parse(JSON.stringify(user));
+  } catch (error) {
+    console.log('error update User Onboarding State', error);
+    console.log('error update User Onboarding State TYPEOF', typeof error);
```
Comment (JazzyGrim) [inline lib/actions/user-actions.ts]: leftover

File: lib/actions/user-actions.ts [hunk 9]
```diff
@@ -56,3 +66,49 @@ export const createNewUser = async ({
...
+  try {
+    await connectToMongoDB();
+    const bla = await User.findByIdAndUpdate({ _id: userId }, data, {
+      new: true,
+    });
+
+    console.log(
+      'BLAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
+      bla
+    );
```
Comment (JazzyGrim) [inline lib/actions/user-actions.ts]: leftover `bla`

File: next.config.mjs [hunk 10]
```diff
@@ -1,4 +1,28 @@
...
+  //   serverComponentsExternalPackages: ['mongoose'], // <-- and this
+  // },
+  // // and the following to enable top-level await support for Webpack
+  // webpack: (config) => {
+  //   config.experiments = {
+  //     topLevelAwait: true,
+  //     layers: true,
+  //   };
+  //   return config;
+  // },
```
Comment (JazzyGrim) [inline next.config.mjs]: leftover

**Review by JazzyGrim — CHANGES_REQUESTED** [review]

File: app/onboarding/page.tsx (hunk 1 above)
Comment (Uros-Bijelic) [reply app/onboarding/page.tsx]: Error file is like a page.tsx like layout.tsx just error.tsx ? thought to tackle this later when i polish the app.

**Pull request merged**